File: 0552-student-attendance-record-ii/0552-student-attendance-record-ii.py
class Solution:
    def __init__(self):
        self.mod = int(1e9 + 7)
        self.dp = [[[-1 for _ in range(3)] for _ in range(2)] for _ in range(100001)]
    
    def sub(self, n, absences, lates):
        if n == 0:
            return 1
        if self.dp[n][absences][lates] != -1:
            return self.dp[n][absences][lates] % self.mod
        
        ans = 0
        
        # Case 1: Adding 'P' (Present)
        ans = (ans + self.sub(n - 1, absences, 0)) % self.mod
        
        # Case 2: Adding 'A' (Absent)
        if absences > 0:
            ans = (ans + self.sub(n - 1, absences - 1, 0)) % self.mod
        
        # Case 3: Adding 'L' (Late)
        if lates < 2:
            ans = (ans + self.sub(n - 1, absences, lates + 1)) % self.mod
        
        self.dp[n][absences][lates] = ans % self.mod
        return ans
    
    def checkRecord(self, n: int) -> int:
        return self.sub(n, 1, 0)

File: 0552-student-attendance-record-ii/0552-student-attendance-record-ii.cpp
class Solution {
private:
    int mod = 1e9+7;
    int dp[100001][2][3];
    //vector<vector<vector<int>>>dp;
    int sub(int n,int abs,int late){
        if(n==0)return 1;
        if(dp[n][abs][late]!=-1)return dp[n][abs][late]%mod;
        int ans = 0;
        int val = 0;
        ans = (ans + sub(n-1,abs,val))%mod;
        if(abs!=0){
            --abs;
            ans = (ans + sub(n-1,abs,val))%mod;
            ++abs;
        }
        if(late<2){
            ++late;
            ans = (ans + sub(n-1,abs,late))%mod;
            --late;
        }
        return dp[n][abs][late]=ans%mod;
    }
public:
    int checkRecord(int n) {
        memset(dp,-1,sizeof(dp));
       // dp.resize(n+1,vector<vector<int>>(2,vector<int>(3,-1)));
        int abs=1,late=0;
        return sub(n,abs,late);
    }
};
